### ISSUE TITLE
CLS breaks when reuse sockets for HTTP fetch

### ISSUE THREAD

**Issue by jchip** [othiym23/node-continuation-local-storage#71]
I have a HTTP fetch that I enable keep alive agent.

something like:
```js
var httpAgent = new require("http").Agent( { keepAlive: true, maxSockets: 1 });
namespace.run( function () {
  var xid = Math.random() * 1000;
  namespace.set( "xid", xid );
  request( "http://www.google.com", { agent: httpAgent }, function (err) {
    assert.equal(namespace.get("xid"), xid);
  });
});
```

Doing that the first time works and CLS context was intact, but the second time when the socket is reused, the CLS context gets the value from the context in the first fetch.

https://gist.github.com/jchip/73c44046408d0e60b96abc3fc0303eb2

**Comment by Jeff-Lewis** [comment]
@jchip In your [gist](https://gist.github.com/jchip/73c44046408d0e60b96abc3fc0303eb2#file-cls-http-agent-break-js), try wrapping superagent's buried EventEmitter. It's ugly but it's working for me.

``` javascript
function httpGetRequest(cb) {

  namespace.bindEmitter(superagent.Request.super_.super_.prototype);

  var r = superagent["get"]("http://www.google.com");

  if (keepAlive) {
    console.log("Keep alive ENABLED, setting http agent");
    r.agent(httpAgent);
  }

  r.end(function (err, res) {
    if (err) {
      cb(err);
    } else {
      console.log("http get status", res.status);
      cb(null, {status: res.status, statusText: res.text, obj: res.body});
    }
  });
}
```

**Comment by jchip** [comment]
nice!  Thanks for figuring that out.

I wonder if it's possible to bind to something lower level in `http` or `socket`, since other fetch libraries have same issue.  I have two other samples that use `wreck` and `node-fetch` in that gist that have this same issue.

**Comment by Qard** [comment]
Yeah, I'm looking into this now and it's proving to be a bit complicated. It seems like `http.request(...)` definitely needs a `ns.bind(...)` on the callback though.

**Comment by Jeff-Lewis** [comment]
In the case of `superagent`, it appears to use its own request object wrapper around `http.request` and encapsulates `http.request`'s events. This wrapper also inherits from `Stream` while using `Stream`'s `EventEmitter` in its public api, not `http.requests`'s `EventEmitter`. 

Thus only cls binding `http`'s `EventEmitter`, I assume, we would always lose context without also wrapping `superagent`'s own `EventEmitter`.

**Comment by jchip** [comment]
@Jeff-Lewis ah... maybe `wreck` and `node-fetch` also have something similar to what `superagent` does.

**Comment by jchip** [comment]
This likely has something to do with it.  

net.Socket.connect is not called when the socket is reused with keep alive and this async wrap would be skipped:

https://github.com/othiym23/async-listener/blob/master/index.js#L62

Any idea on how to mitigate that?

**Comment by jchip** [comment]
This seems to do the trick:
```js
wrap(net.Socket.prototype, 'unref', function (original) {
  return function() {
    if (this._saveOnRead) {
      this._handle.onread = this._saveOnRead;
      delete this._saveOnRead;
    }
    original.apply(this);
  }
});


wrap(net.Socket.prototype, 'ref', function (original) {
  return function() {
    if (this._handle) {
      this._saveOnRead = this._handle.onread;
      this._handle.onread = wrapCallback(this._handle.onread);
    }
    original.apply(this);
  }
});

wrap(net.Socket.prototype, 'connect', function (original) {
  return function () {
    var args = net._normalizeConnectArgs(arguments);
    if (args[1]) args[1] = wrapCallback(args[1]);
    var result = original.apply(this, args);
    if (this._handle) {
      this._saveOnRead = this._handle.onread;
      this._handle.onread = wrapCallback(this._handle.onread);
    }
    return result;
  };
});
```

**Comment by Jeff-Lewis** [comment]
FWIW - I made of version of CLS using [AsyncWrap](https://github.com/nodejs/node-eps/blob/async-wrap-ep/XXX-asyncwrap-api.md) (and [async-hook](https://github.com/AndreasMadsen/async-hook)) instead of [async-listener](https://github.com/othiym23/async-listener) called [cls-hooked](https://github.com/Jeff-Lewis/cls-hooked). In my tests, it worked with Agent without the patch to net.Socket. I guess (hope) this is one example AsyncWrap removing the need for many of the core node patches done in async-listener.

**Comment by jchip** [comment]
That is awesome!!!  Hopefully AsyncWrap will be official soon.  I will definitely try your cls-hooked sometime.  Thanks.

**Comment by jchip** [comment]
@Jeff-Lewis I tried cls-hooked with https://gist.github.com/jchip/73c44046408d0e60b96abc3fc0303eb2.  It breaks the same way before the socket patch for async-listener.  Is there anything else you did with using cls-hooked to make it work?

**Comment by jchip** [comment]
@Jeff-Lewis I tried http-agent-break.test.js with continuation-local-storage and async-listener@0.6.0 (before the agent patch) and it pass also, while https://gist.github.com/jchip/73c44046408d0e60b96abc3fc0303eb2 fails.

When I was debugging this, it was also very tricky.  It worked in some way and fails in some other way.  It took me a while to find the approach in my test to cause it to fail consistently.  I am not sure what's the diff in your test with a quick look though.

I think using async wrap also requires the http.Agent patch.
